src/MOI/MOI_wrapper.jl

```diff
@@ -770,10 +777,15 @@ _sense_and_rhs(s::MOI.EqualTo{Float64}) = (Cchar('E'), s.value)
 
 # Short-cuts to return the VariableInfo associated with an index.
 function _info(model::Optimizer, key::MOI.VariableIndex)
-    if haskey(model.variable_info, key)
-        return model.variable_info[key]
+    if !haskey(model.variable_info, key)
+        _invalid_index(key)
     end
+    return model.variable_info[key]
+end
+
+function _invalid_index(key::MOI.VariableIndex)
     throw(MOI.InvalidIndex(key))
+    return nothing
```

Comment:
What was the motivation for this change? I generally try to avoid introducing small one-line functions which are only used once.

Reply:
It's not clear to me why, but julia can't infer the type of throw(MOI.InvalidIndex(key)) (as showed #179 %8) , even though the types of all the structures involved are well defined. So, the idea is to remove this line from the function, so the instability only occurs at runtime when it gives an error, which doesn't seem so serious to me.
Maybe @joaquimg can give more details

Reply:
Ah. It's because the `throw` was returning. I just pushed a commit which swaps the returns. That should type stable. Note that it might flag `throw` as being type unstable, but that doesn't matter because the value won't be used unless there is an error. The return type of the function _will_ be stable, which is what matters.

Reply:
Did you test this? I don't know if you have the Xpress license to do this.
But it seems to me that the problem persists.
![WhatsApp Image 2022-09-14 at 23 40 07](https://user-images.githubusercontent.com/77932135/190301036-c9de3b1b-37fe-4727-b72a-9df116e4ac28.jpeg)

Reply:
I didn't test this because I don't have a license, but this is what I meant by

> Note that it might flag throw as being type unstable, but that doesn't matter because the value won't be used unless there is an error. The return type of the function will be stable, which is what matters.

You really care about the `Body::Xpress.VariableInfo` part of `@code_warntype`. That's the thing that matters.

Reply:
You can actually see that the `Any` is a quirk of the compiler. It doesn't infer anything that is passed to `throw`, because the cost of `throw` is far larger than the benefit of inference.
```Julia
julia> function foo(x::MOI.VariableIndex)
       y = MOI.InvalidIndex(x)
       throw(y)
       end
foo (generic function with 1 method)

julia> @code_warntype foo(MOI.VariableIndex(1))
Variables
  #self#::Core.Const(foo)
  x::MathOptInterface.VariableIndex
  y::Any

Body::Union{}
1 ─ %1 = MathOptInterface.InvalidIndex::Core.Const(MathOptInterface.InvalidIndex)
│        (y = (%1)(x))
│        Main.throw(y)
└──      Core.Const(:(return %3))

julia> function foo(x::MOI.VariableIndex)
       y = MOI.InvalidIndex(x)
       #throw(y)
       end
foo (generic function with 1 method)

julia> @code_warntype foo(MOI.VariableIndex(1))
Variables
  #self#::Core.Const(foo)
  x::MathOptInterface.VariableIndex
  y::MathOptInterface.InvalidIndex{MathOptInterface.VariableIndex}

Body::MathOptInterface.InvalidIndex{MathOptInterface.VariableIndex}
1 ─ %1 = MathOptInterface.InvalidIndex::Core.Const(MathOptInterface.InvalidIndex)
│   %2 = (%1)(x)::MathOptInterface.InvalidIndex{MathOptInterface.VariableIndex}
│        (y = %2)
└──      return %2
```

Reply:
Actually, the problem seems to be in `MOI.InvalidIndex(key)`, right?
I don't think `throw` is causing this.

Reply:
I thought the same @odow, but we used the profiler and the call to this function was both marker as runtime dispatch and it was taking soo much more time than expected.
@Thiago-NovaesB can send the profiler flamegraph.

Reply:
Even with the latest version?

Reply:
Current version was the first thing we tried to fix the issue.
We can double check the flamegraphs, if that is indeed true, it might be impacting other stuff. So would be good to be sure.

Reply:
By current, I mean latest version. Your code suggestion.

Reply:
@odow your observation is correct, the instability in the flamegraph was caused by the cleverdict, `::Any` is not a problem.
So, this version with your last commit looks good to me, the tests pass and the type instability has been removed.